File: Data Structures & Algorithms/jump-game-vii/submission-1.cpp
class Solution {
public:
    vector<int>dp;

    bool canReach(string s, int minJump, int maxJump) {
        int n = s.length();
        dp.resize(n, -1);

        dp[n-1] = 1;

        if(s[n-1] == '1')
            return false;

        return dfs(0, s, minJump, maxJump, n);
    }
    bool dfs(int i, string &s, int& minJump, int& maxJump, int n){
        if(dp[i] != -1)
            return dp[i];

        dp[i] = 0;

        for(int j= i+minJump; j<= min(n-1, i+maxJump); j++){
            if(s[j] == '0' && dfs(j, s, minJump, maxJump, n)){
                dp[i] = 1;
                break;
            }
        }

        return dp[i];
    }
};

File: Data Structures & Algorithms/islands-and-treasure/submission-1.java
class Solution {

    void dfs(int[][] grid, int i, int j, int dist){
        if(i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == -1 || grid[i][j] < dist)
            return;

        grid[i][j] = dist;

        dfs(grid, i+1, j, dist+1);
        dfs(grid, i-1, j, dist+1);
        dfs(grid, i, j+1, dist+1);
        dfs(grid, i, j-1, dist+1);

    }

    public void islandsAndTreasure(int[][] grid) {
        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[i].length;j++){
                if(grid[i][j] == 0)
                    dfs(grid, i, j, 0);
            }
        }
    }
}
